src/vector_transport.jl

```diff
@@ -0,0 +1,636 @@
+    return vector_transport_along(M, p, X, c, ParallelTransport())
+end
+function vector_transport_along(M::Manifold, p, X, c, m::AbstractVectorTransportMethod)
+    Y = allocate_result(M, vector_transport_along, X, p)
+    vector_transport_along!(M, Y, p, X, c, m)
+    return Y
+end
+
+"""
+    vector_transport_along!(M::Manifold, Y, p, X, c)
+    vector_transport_along!(M::Manifold, Y, p, X, c, method::AbstractVectorTransportMethod)
+
+Transport a vector `X` from the tangent space at a point `p` on the [`Manifold`](@ref) `M`
+along the curve `c` such that `c(0)` is equal to `p` to the point `c(1)` using the `method`,
+which defaults to [`ParallelTransport`](@ref). The result is saved to `Y`.
+"""
+function vector_transport_along!(M::Manifold, Y, p, X, c)
+    return vector_transport_along!(M, Y, p, X, c, ParallelTransport())
+end
+function vector_transport_along!(
+    M::Manifold,
+    Y,
+    p,
+    X,
+    c,
+    method::AbstractVectorTransportMethod,
+)
+    return error(manifold_function_not_implemented_message(
+        M,
+        vector_transport_along!,
+        M,
+        Y,
+        p,
+        X,
+        c,
+        method,
+    ))
+end
+@doc raw"""
+    vector_transport_along!(
+        M::Manifold,
+        Y,
+        p,
+        X,
+        c::AbstractVector,
+        method::AbstractVectorTransportMethod
+    ) where {T}
+
+Compute the vector transport along a discretized curve `c` using an
+[`AbstractVectorTransportMethod`](@ref) `method` succesively along the sampled curve.
+"""
+function vector_transport_along!(
+    M::Manifold,
+    Y,
+    p,
+    X,
+    c::AbstractVector,
+    method::AbstractVectorTransportMethod,
+)
+    n = length(c)
+    if n == 0
+        copyto!(Y, X)
+    else
+        # we shouldn't assume that vector_transport_to! works when both input and output
+        # vectors are the same object
+        Y2 = allocate(X)
+        vector_transport_to!(M, Y2, p, X, c[1], method)
+        for i in 1:(length(c) - 1)
+            vector_transport_to!(M, Y, c[i], Y2, c[i + 1], method)
+            copyto!(Y2, Y)
+        end
+    end
+    return Y
+end
+@doc raw"""
+    function vector_transport_along!(
+        M::Manifold,
+        Y,
+        p,
+        X,
+        c::AbstractVector,
+        method::PoleLadderTransport
+    )
+
+Compute the vector transport along a discretized curve using
+[`PoleLadderTransport`](@ref) succesively along the sampled curve.
+This method is avoiding additional allocations as well as inner exp/log by performing all
+ladder steps on the manifold and only computing one tangent vector in the end.
+"""
+function vector_transport_along!(
+    M::Manifold,
+    Y,
+    p,
+    X,
+    c::AbstractVector,
+    method::PoleLadderTransport,
+)
+    d = exp(M, p, X)
+    m = p
+    clen = length(c)
+    for i in 1:(clen - 1)
+        # precompute mid point inplace
+        ci = c[i]
+        cip1 = c[i + 1]
+        log!(M, Y, ci, cip1)
+        exp!(M, m, ci, Y / 2)
```

Comment:
Do we need to use `log!` and `exp!` here instead of retraction and inverse retraction from `method`?

Reply:
Similarly in other cases below.

Reply:
I think so since we need the mid point.

Reply:
OK, in this case it makes sense, but what about other `exp`s and `log`s? for example in lines 409, 431, 458, 480?

Reply:
BTW, maybe we should add `mid_point` and `mid_point!` that you have in Manopt.jl so that it can be used here?

Reply:
I would actually be in favour of that, yes :)

For the remaining lines:
* 409, 431 – well since in between we use retractions to move around, we could do that there too,
* similarly 458 and 480
* but... 579, 586, 602, and 609 should then be adapted, too.

For all of those (even the retracts we already use) I am not sure about theory, but yes, _if_ we leave log/exp and generalise we could do for all non-mids.

Reply:
OK, I'll add `mid_point` and adapt these lines.

Reply:
I think it should work correctly after the last commit.